### PR TITLE
WMDP-123: Add linting, logging, and several other core utils for API

### PR DESCRIPTION
**Pull request by chouinar**
## Ticket

https://wicmtdp.atlassian.net/browse/WMDP-123

## Changes
Added the following linters:
* Spectral (openapi validation)
* isort & black (formatting)
* flake8 (bad practice catcher)
* mypy (type checking)

Makefile to help coordinate the linters.

Logging so we can see what is going on when it runs.

Various other utilities that I just grabbed along the way that'll be needed:
* Response object creation
* API auto-reload on code change

## Context for reviewers

If you've worked on some of our other projects, you probably recognize a lot of these patterns.

I shuffled a few bits around like how the app is initialized to follow existing formats (ie. the app creation is shoved in a function so we can use it for testing when I do that next).

I created a basic API with the relevant classes mostly to validate the type checking + spectral.

## Testing

Here is the current output of running `make lint`. Fixed all the code issues, and when the API actually gets built out, will fix these minor spectral warnings.
Screen Shot 2022-08-29 at 10 26 33 AM

Here's what the logging looks like including an auto-reload.
Screen Shot 2022-08-29 at 10 29 28 AM

### CODE REVIEW

File: app/Dockerfile [hunk 1]
```diff
@@ -7,10 +7,18 @@ RUN apt update \
...
+
 # Set the application working directory.
 WORKDIR /srv
+
+COPY pyproject.toml poetry.lock ./
+RUN poetry config virtualenvs.in-project false && poetry env use python
+RUN poetry install --no-root
+
+# Copy application files.
+COPY . /srv
```
Comment (rocketnova) [inline app/Dockerfile:20]: Very nice

File: app/Makefile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+lint-mypy:
+	$(PY_RUN_CMD) mypy --show-error-codes api
+
+lint-poetry-version: ## Check poetry version
+	grep --quiet 'lock-version = "1.1"' poetry.lock
+
+lint-spectral:
+	docker run --rm --tty --cap-drop=ALL --network=none --read-only --volume=$(PWD):/tmp:ro \
+		stoplight/spectral:6 lint /tmp/openapi.yml --ruleset /tmp/.spectral.yaml
```
Comment (rocketnova) [inline app/Makefile]: I know I mentioned a slight reluctance around Make, but this is a very nice Makefile! Thanks for this!

File: app/api/__main__.py [hunk 3]
```diff
@@ -5,20 +5,27 @@
...
 
-def get_project_root_dir() -> str:
-    return os.path.join(os.path.dirname(__file__), "..")
+    if is_local:
+        # If python files are changed, the app will auto-reload
+        # Note this doesn't have the OpenAPI yaml file configured at the moment
+        app.run(port=8080, use_reloader=True, reloader_type="stat")
+    else:
+        # Don't enable the reloader if non-local
+        app.run(port=8080)
```
Comment (rocketnova) [inline app/api/__main__.py:28]: Very nice!

File: app/openapi.yml [hunk 4]
```diff
@@ -20,3 +20,128 @@ paths:
...
+
+    ValidationErrorDetail:
+      type: object
+      properties:
+        type:
+          example: "required"
+          type: string
+        message:
+          example: "request.first_name is required."
+          description: "Somewhat user-friendly description of the error. May contain PII."
```
Comment (rocketnova) [inline app/openapi.yml:118]: Good flag. We should keep PII out of the errors and logs.

File: app/api/util/response.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    errors: Optional[list[ValidationErrorDetail]] = None
+
+    def to_dict(self) -> dict[str, Any]:
+        return exclude_none(asdict(self))
+
+    def to_api_response(self) -> flask.Response:
+        return flask.make_response(flask.jsonify(self.to_dict()), self.status_code)
+
+
+def exclude_none(obj):
```
Comment (rocketnova) [inline app/api/util/response.py:33]: Neat!

**Review by rocketnova — APPROVED** [review]
LGTM! Thank you! 🚢 🎉 

Like we chatted about, let's add the flexibility to run the make commands inside the docker container as well.

**Comment by chouinar** [comment]
@rocketnova - FYI I updated the Makefile to let you specify whether you want to run inside the Docker container or not for python scripts. Just requires flipping an env var to switch.

Also brought in some CI configurational differences on the linting, although didn't test them.

Will merge if no further questions tomorrow morning.

**Review by rocketnova — APPROVED** [review]
Looks great! Thanks for the speedy turnaround!

**Pull request merged**